### PR TITLE
[feat] add visualization API and notebook

### PR DESCRIPTION
**Pull request by brettallenyo**
Add a visualization API for the frcnn
features, which takes in image and feature
paths and uses the huggingface visualization
tools to show the image.

Also added a jupyter notebook that will display
the images accordingly if the features and images
are in the correct place.

### CODE REVIEW

**Comment by brettallenyo** [comment]
Screen Shot 2021-02-05 at 12 29 15 PM

The notebook looks like this in case people were wondering

File: tools/scripts/features/frcnn/ImageVisualizer.ipynb
Comment (apsdehal) [inline tools/scripts/features/frcnn/ImageVisualizer.ipynb:1]: We usually land notebooks in `notebooks` branch and do not land any ipython code in master.

Comment (brettallenyo) [reply tools/scripts/features/frcnn/ImageVisualizer.ipynb:1]: Ah thats great to know thank you, I will adjust

File: tools/scripts/features/frcnn/utils.py
Comment (vedanuj) [inline tools/scripts/features/frcnn/utils.py:1]: Is this file required for this PR? Also this has lot of unnecessary code. Let's remove them.

**Review by vedanuj — CHANGES_REQUESTED** [review]
This is great. Excited to see the notebook running. Please restructure the code as per my comments. Also as @apsdehal noted, let's not add the ipython notebook in this PR. We can add it later to the notebooks branch. 

It will be good to keep this PR only about the visualization API that you are adding to the utils. It will be good to write some tests to check working of this API as well.

**Comment by vedanuj** [comment]
> What should I use for the image and feature names to abstract the notebook for all users

The notebook can add the full functionality of extracting the features and displaying the bounding boxes. If the user inputs an image, the notebook will have step by step instructions of extracting the features (you might have to link to the feature extraction apis that you have implemented in the other PR #764), then visualize the boxes that were extracted. 

> Should I include the training and inference command line code so the notebook is a one-stop-shop for everything I have done so far

Training and inference doesn't need to be coupled with this. This notebook can just be about feature extraction and visualization.

> Should I upload the images and features so it works everywhere even without the model training

Not sure I understand the question properly. We will need to extract the features using your script in PR #764 for COCO and Visual Genome datasets. We will store the features locally in FAIR Cluster to train the multimodal models with these extracted features. These models will later be used for end to end inference once we add your frcnn script as an encoder.

File: mmf/utils/features/visualizing_image.py
Comment (vedanuj) [inline mmf/utils/features/visualizing_image.py:21]: Let's try to use PIL image library instead of cv2 if possible. Adding cv2 to utils will require additional dependency to mmf core which we should try to avoid if possible.

**Review by vedanuj — CHANGES_REQUESTED** [review]
Some tests are failing. Those are because of adding additional dependencies like cv2. Let's see if we can use PIL reader instead. 

Also rebase this on master and then import this to Phabricator to check if internal tests pass.

File: mmf/utils/features/visualizing_image.py
Comment (ytsheng) [inline mmf/utils/features/visualizing_image.py:50]: cvtColor is doing exactly `img = img[:, :, ::-1]`, so ifyou do cvtColor(cv2.COLOR_BGR2RGB) twice I believe you will get back the original image. So the logic here seems a bit weird.

**Review by ytsheng — COMMENTED** [review]
Looks good to me. cv2 is pretty heavy - PIL seems better. :)

File: mmf/utils/features/visualizing_image.py
Comment (brettallenyo) [reply mmf/utils/features/visualizing_image.py]: I agree, I think it is some weird cv2 behavior that I just dropped in accordance with Vedanuj's comment about removing that dependency

File: mmf/utils/features/visualizing_image.py
Comment (apsdehal) [inline mmf/utils/features/visualizing_image.py:34]: I would suggest moving to `mmf/utils/download.py`.

File: mmf/utils/features/visualizing_image.py
Comment (apsdehal) [inline mmf/utils/features/visualizing_image.py:50]: Do we plan to add typings to this? or some sort of tests?

File: mmf/utils/features/visualizing_image.py
Comment (apsdehal) [inline mmf/utils/features/visualizing_image.py:1]: Is there a reason this can't also go into utils/visualize.py?

File: mmf/utils/visualize.py
Comment (apsdehal) [inline mmf/utils/visualize.py:54]: Isn't this technically same as `img_tensorize`?

File: mmf/utils/visualize.py
Comment (apsdehal) [inline mmf/utils/visualize.py:56]:
```suggestion
    output_dict = np.load(features_path, allow_pickle=True).item()
```

File: mmf/utils/visualize.py
Comment (brettallenyo) [reply mmf/utils/visualize.py:54]: Actually yes, I am realizing I can actually just remove that line and pass in the image path directly

File: mmf/utils/features/visualizing_image.py
Comment (brettallenyo) [reply mmf/utils/features/visualizing_image.py:1]: I could definitely put it there, I just did not want to overcrowd that file with feature-displaying-specific code, do you think I should move it?

File: mmf/utils/features/visualizing_image.py
Comment (vedanuj) [reply mmf/utils/features/visualizing_image.py:65]: @brettallenyo Please add typings to all the functions in this file.

File: mmf/utils/visualize.py
Comment (vedanuj) [inline mmf/utils/visualize.py:53]: Add typings here as well

Comment (vedanuj) [reply mmf/utils/visualize.py:53]: Might be good to write a test for this method. One way to test is make sure the input and output image arrays are same.

**Review by vedanuj — APPROVED** [review]
Looks good to me. Thanks for adding the test. Please rebase the branch before landing.

**Pull request closed without merging**
